KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+- `Registered - 1`: Retirees and Newbies are registered.
+- `Approved - 2`: All retirees approved by msg.sender
+- `Finalized - 3`: Rebalance executed and finalized.
+
+#### Life Cycle
+
+The contract status should follow the ENUM order above during status transition. The only way to go to previous state is by calling Reset() function.
+
+![](../assets/kip-103/lifecycle.png)
+
+Status transition
+
+- Initialized → Registered → Approved → Finalized ✅
+- Initialized → Registered → Finalized ❌
+- Initialized → Approved ❌
+- Initialized → Finalized ❌
+
+#### Structs
+
+The smart contract will have the following structs:
+
+- `Retired`: to represent the details of retired and approver addresses.
+- `Newbie`: to represent newbies and their fund allocation.
+
+#### Storage
+
+The smart contract will have the following storage variables:
+
+- `retirees`: array of `Retired` struct.
+- `newbies`: array of `Newbie` struct.
+- `status`: current status of the contract.
+- `rebalanceBlockNumber`: the target block number of the execution of rebalancing.
+- `memo`: result of the treasury fund rebalance. eg : `Treasury Rebalanced Successfully`
+
+#### Modifiers
+
+The smart contract will have the following modifier:
+
+- `onlyAtStatus`: to restrict access to certain functions based on the current status of the contract. If the status is not the same with the given status, it reverts.
+
+#### Constructor
+
+The smart contract will have the following constructor:
+
+- `constructor`: to initialize the contract with the target block number of the execution of rebalance.
+
+#### State Changing Functions
+
+The smart contract will have the following state changing functions:
+
+- `registerRetired`: to register retired details. Retired stores the retired address and approvers
+- `removeNewbie`: to remove retired details from the array.
+- `registerRetired`: to register newbie address and its fund distribution.
+- `removeNewbie`: to remove newbie details from the array.
+- `finalizeRegistration`: sets the status to Registered only executed by owner. After this stage, registrations will be restricted.
+- `approve`: to approve a retiredAddress by the admin.
+- `finalizeApproval`: sets the status to Approved. After this stage, approvals will be restricted.
+  </br>
+  **Conditions**
+  - every retired must be approved
+  - min required admin’s approval is done for retired contract address.
+  - sum of retired’s balance is greater than treasury amount
+    </br>
+- `reset`: resets all storage values to empty objects except rebalanceBlockNumber. It can only be called during Initialization, Registration and Approved status.
+- `finalizeContract`: to record the execution result of the rebalance and finalize the contract. The storage data cannot be modified after this stage.
+
+#### Fallback Function
+
+The smart contract will have a fallback function to revert any payments. Since its a treasury contract it should not accept any payments nor it should allow any withdrawal.
```

Comment:
```suggestion
The smart contract will have a fallback function to revert any payments. Since its a treasury contract, it should not accept any payments nor it should allow any withdrawal.
```